### PR TITLE
Remove hardcode

### PR DESCRIPTION
**Pull request by LeonardoTredese**
Plz review

### CODE REVIEW

File: dataset/dataset.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 import json
 import torch
 
-class Cityscapes(Dataset):
+class SegmentationDataset(Dataset):
```
Comment (Giost) [inline dataset/dataset.py:9]: Perché hai cambiato nome? Mi sembra sia più chiaro con Cityscapes, dato che è pensato per quel dataset e visto che dovremmo poi crearne un'altra per GTA5, inoltre anche nella repo originale fa così (https://github.com/ooooverflow/BiSeNet/blob/master/dataset/CamVid.py)

File: run.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+python train.py  --num_epochs 50 \
+        --learning_rate 2.5e-2 \
+        --data ../data/ \
+        --num_classes 19 \
+        --cuda 0 \
+		--validation_step 10 \
+		--num_workers 8 \
+        --batch_size 32 \
```
Comment (Giost) [inline run.sh]: NIT: nel paper usano batch da 16

File: dataset/dataset.py [hunk 3]
```diff
@@ -51,7 +51,7 @@ def __getitem__(self, index):
         img_path = os.path.join(self.path, 'images', img_file_name)
         lbl_path = os.path.join(self.path, 'labels', lbl_file_name)
         img = self.img_transformer(Image.open(img_path)).to(torch.float32).to(self.device)
-        lbl = self.lbl_transformer(Image.open(lbl_path)).to(torch.float32).to(self.device)[0]
+        lbl = self.lbl_transformer(Image.open(lbl_path)).to(torch.float32).long().to(self.device)
```
Comment (Giost) [inline dataset/dataset.py]: Possiamo risparmiarci di convertire le labels in torch.float32, io avevo provato e funziona anche senza

File: train.py [hunk 4]
```diff
@@ -41,18 +41,18 @@ def val(args, model, dataloader, final_test):
...
-    # label_info = get_label_info(csv_path)
     with torch.no_grad():
         model.eval()
         precision_record = []
         hist = np.zeros((args.num_classes, args.num_classes))
         for i, (data, label) in enumerate(dataloader):
             label = label.type(torch.LongTensor)
-            data = data.cuda()
-            label = label.long().cuda()
+            data = data
```
Comment (Giost) [inline train.py]: Non è più utile, da rimuovere

File: train.py [hunk 5]
```diff
@@ -41,18 +41,18 @@ def val(args, model, dataloader, final_test):
...
+            label = label.long()
```
Comment (Giost) [inline train.py]: Ho visto che hai aggiunto la conversione in long già nella dataset class, quindi potrebbe essere inutile questa riga

File: train.py [hunk 6]
```diff
@@ -41,18 +41,18 @@ def val(args, model, dataloader, final_test):
...
 
             # get RGB predict image
             predict = model(data).squeeze()
+            predict = predict.squeeze()
```
Comment (Giost) [inline train.py]: squeeze viene già fatto nella riga sopra, serve a qualcosa farlo di nuovo?

File: train.py [hunk 7]
```diff
@@ -98,24 +98,22 @@ def val(args, model, dataloader, final_test):
...
         lr = poly_lr_scheduler(optimizer, args.learning_rate, iter=epoch, max_iter=args.num_epochs)
         model.train()
         tq = tqdm(total=len(dataloader_train) * args.batch_size)
         tq.set_description('epoch %d, lr %f' % (epoch, lr))
         loss_record = []
         for i, (data, label) in enumerate(dataloader_train):
-            data = data.cuda()
-            label = label.long().cuda()
+            data = data
+            label = label.long()
```
Comment (Giost) [inline train.py]: Come nel commento sopra

File: train.py [hunk 8]
```diff
@@ -139,31 +137,27 @@ def train(args, model, optimizer, dataloader_train, dataloader_val):
...
-            torch.save(model.module.state_dict(),
-                       os.path.join(args.save_model_path, 'latest_dice_loss.pth'))
-
-        if epoch % args.validation_step == 0 and epoch != 0:
+        is_right_iteration = lambda i, step: (i % step) == (step - 1)
+        if is_right_iteration(epoch, args.checkpoint_step) and args.model_file_name is not None:
+            model_path = os.path.join(args.saved_models_path, args.model_file_name)
+            torch.save(model.module.state_dict(), model_path) 
+
+        if is_right_iteration(epoch, args.validation_step) and args.model_file_name is not None:
```
Comment (Giost) [inline train.py]: Forse sarebbe meglio fare la validation anche se non è specificato un modello, magari quel controllo lo puoi spostare nell'if sotto

**Review by Giost — COMMENTED** [review]
Ho lasciato alcuni commenti

File: dataset/dataset.py (hunk 1 above)
Comment (LeonardoTredese) [reply dataset/dataset.py:9]: Ho cambiato il nome perchè il preprocessing che facciamo mi sembra compatibile con entrambi i dataset
